### PR TITLE
Introduce generated client for `core/v1` resources

### PR DESCRIPTION
**Pull request by fossedihelm**
### What this PR does
Before this PR:
We are using the generated client for some resources, such as `clone` or `migration`, while we are using
the custom, manually maintained, for the core resources (VM, VMI...)

After this PR:
Convert using the generated client instead of the custom one.
This is the path will be followed:
1. Introduction of the generated client.
2. Redirect common api calls to the generated client
leaving the custom struct as extended wrapper.
3. Drop the custom wrapper structs, moving to the generated
resources only.

**NB: This PR covers just the first two steps, since otherwise the changes will result
too big.***

Fixes #

### Why we need it and why it was done in this way
The generated client has the advantage that the various methods will have the same standard signature. Currently, the method's signature between `Get`/`Create`/`Update` etc.. are different between VM/VMI/etc..:
- sometimes a context is required, sometimes no
- sometimes the options are required, sometimes no
- sometimes a pointer is used as options, sometimes directly the struct

Also, alongside the generated client, the `client-gen` command generates the fakeclient that can be used in unit tests.
The fakeclient has the benefit, in unit tests, of simulating the real client and recording the changes, getting rid of mocking API requests (that can create misalignment to the reality) e.g. https://github.com/kubevirt/kubevirt/pull/10908

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

**Comment by Barakmor1** [comment]
Hello @fossedihelm ,

I noticed in the commit message that you’ve outlined the task as consisting of three steps. In your commits, you’ve labeled them as “step 1” and “step 2.” However, I’m a bit confused because there seems to be a third step mentioned, but I only see two steps addressed in the commits. Could you please clarify if the second and third steps are included within the same commits, or am i'm missing something?

**Comment by Barakmor1** [comment]
> > Hello @fossedihelm ,
> > I noticed in the commit message that you’ve outlined the task as consisting of three steps. In your commits, you’ve labeled them as “step 1” and “step 2.” However, I’m a bit confused because there seems to be a third step mentioned, but I only see two steps addressed in the commits. Could you please clarify if the second and third steps are included within the same commits, or am i missing something?
> 
> Hey @Barakmor1 sorry about that, as I mentioned in the PR description, this PR only covers the first two steps (Now I have highlighted that part). I thought having the full overview of the journey in the commits was good. Hope this helps. Thank you again!

Thanks for clarifying

File: staging/src/kubevirt.io/client-go/kubecli/migration.go [hunk 1]
```diff
@@ -148,12 +102,12 @@ func (v *migration) PatchStatus(name string, pt types.PatchType, data []byte) (r
...
+func (o *migration) UpdateStatus(vmi *v1.VirtualMachineInstanceMigration) (result *v1.VirtualMachineInstanceMigration, err error) {
 	result = &v1.VirtualMachineInstanceMigration{}
-	err = v.restClient.Put().
+	err = o.restClient.Put().
 		Name(vmi.ObjectMeta.Name).
-		Namespace(v.namespace).
-		Resource(v.resource).
+		Namespace(o.namespace).
+		Resource(o.resource).
 		SubResource("status").
```
Comment (Barakmor1) [inline staging/src/kubevirt.io/client-go/kubecli/migration.go:111]: I think you forgot to use ` o.VirtualMachineInstanceMigrationInterface.UpdateStatus` here

**Comment by Barakmor1** [comment]
I would also remove the 3 step from the commit message or at least mention that it will be handled in the future so it will be clear just from reading the commit messages.

File: staging/src/kubevirt.io/client-go/kubecli/migration.go (hunk 1 above)
Comment (fossedihelm) [reply staging/src/kubevirt.io/client-go/kubecli/migration.go]: Nice catch! You're right! Thanks for catching this! Done

**Comment by fossedihelm** [comment]
> I would also remove the 3 step from the commit message or at least mention that it will be handled in the future so it will be clear just from reading the commit messages.

Thank you! I dropped the mention of the 3rd step in the commits.

**Comment by Barakmor1** [comment]
Thank you !
/lgtm

**Review by orelmisan — COMMENTED** [review]
Thank you for the PR @fossedihelm 
I am looking forward to be able to use the auto-generated fake clients in tests.

File: staging/src/kubevirt.io/client-go/kubecli/vmi.go [hunk 2]
```diff
@@ -289,26 +290,13 @@ func (v *vmis) Unpause(ctx context.Context, name string, unpauseOptions *v1.Unpa
...
 func (v *vmis) Get(ctx context.Context, name string, options *k8smetav1.GetOptions) (vmi *v1.VirtualMachineInstance, err error) {
-	vmi = &v1.VirtualMachineInstance{}
-	err = v.restClient.Get().
-		Resource(v.resource).
-		Namespace(v.namespace).
-		Name(name).
-		VersionedParams(options, scheme.ParameterCodec).
-		Do(ctx).
-		Into(vmi)
+	vmi, err = v.VirtualMachineInstanceInterface.Get(ctx, name, *options)
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply staging/src/kubevirt.io/client-go/kubecli/vmi.go:293]: This information is added just before serialization by https://github.com/kubernetes/apimachinery/blob/0f2e9357997ff027c5ec6b9a8adb4a38c05fa02e/pkg/runtime/helper.go#L222

Based on information from (for example) https://github.com/kubernetes/client-go/blob/8c4efe8d079e405329f314fb789a41ac6af101dc/kubernetes/typed/apps/v1/apps_client.go#L107

[the remark this answers is not in the document]
Comment (fossedihelm) [reply staging/src/kubevirt.io/client-go/kubecli/vmi.go:293]: Thank you for the information. 
Consider that this part will be totally dropped in a quite already prepared patch: https://github.com/fossedihelm/kubevirt/commit/fd5f1fa237ce7d59cfc1a74f5573409f30b86e57 :)

File: staging/src/kubevirt.io/client-go/kubecli/kubevirt.go [hunk 3]
```diff
@@ -84,6 +85,7 @@ type KubevirtClient interface {
 	GuestfsVersion() *GuestfsVersion
 	RestClient() *rest.RESTClient
 	GeneratedKubeVirtClient() generatedclient.Interface
+	CoreClient() *kvcorev1client.Clientset
```
Comment (xpivarc) [inline staging/src/kubevirt.io/client-go/kubecli/kubevirt.go:88]: Why do we need this?

File: staging/src/kubevirt.io/client-go/kubecli/replicaset.go [hunk 4]
```diff
@@ -156,15 +121,7 @@ func (v *rc) PatchStatus(name string, pt types.PatchType, data []byte) (result *
 }
 
 func (v *rc) UpdateStatus(vmi *v1.VirtualMachineInstanceReplicaSet) (result *v1.VirtualMachineInstanceReplicaSet, err error) {
-	result = &v1.VirtualMachineInstanceReplicaSet{}
```
Comment (xpivarc) [inline staging/src/kubevirt.io/client-go/kubecli/replicaset.go:159]: Missed `PatchStatus` above?

File: staging/src/kubevirt.io/client-go/kubecli/kv.go [hunk 5]
```diff
@@ -28,118 +28,73 @@ import (
...
 func (o *kv) List(options *k8smetav1.ListOptions) (*v1.KubeVirtList, error) {
-	newKvList := &v1.KubeVirtList{}
-	err := o.restClient.Get().
-		Resource(o.resource).
-		Namespace(o.namespace).
-		VersionedParams(options, scheme.ParameterCodec).
-		Do(context.Background()).
-		Into(newKvList)
-
+	newKvList, err := o.KubeVirtInterface.List(context.Background(), *options)
```
Comment (xpivarc) [inline staging/src/kubevirt.io/client-go/kubecli/kv.go:81]: what if `options` is `nil`?

**Review by xpivarc — COMMENTED** [review]
/approve
/hold

File: staging/src/kubevirt.io/client-go/kubecli/vmi.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/kubecli/vmi.go:293]: >Honestly, I am not sure if this explicitly set is required or not. (@xpivarc do you have any insight?)

I know that this did not work in the past, my memory is bad at this time so, I can't say why...
What I can see is that the request contains this information when I try `kubectl get vmi <name> -o yaml --v 9`

File: staging/src/kubevirt.io/client-go/kubecli/kubevirt.go (hunk 3 above)
Comment (fossedihelm) [reply staging/src/kubevirt.io/client-go/kubecli/kubevirt.go]: Leftover :man_facepalming: Removed entirely the commit

File: staging/src/kubevirt.io/client-go/kubecli/replicaset.go (hunk 4 above)
Comment (fossedihelm) [reply staging/src/kubevirt.io/client-go/kubecli/replicaset.go:159]: There isn't a `PatchStatus` method in the replicaset interface :)

**Comment by fossedihelm** [comment]
[Changes](https://github.com/kubevirt/kubevirt/compare/64f677c9087208882a45a23b9a143012d4f722f8..d0789bd5ff2366af179e9d37b1f4dac20ad8396b) Addressed possible nil pointer reference.

File: staging/src/kubevirt.io/client-go/kubecli/replicaset.go (hunk 4 above)
Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/kubecli/replicaset.go:159]: I think you want to use `subresources` argument of the `Patch`.

Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/kubecli/replicaset.go:159]: @fossedihelm

Comment (fossedihelm) [reply staging/src/kubevirt.io/client-go/kubecli/replicaset.go:159]: This is a good idea! Thank you!

**Comment by fossedihelm** [comment]
[Changes](https://github.com/kubevirt/kubevirt/compare/d0789bd5ff2366af179e9d37b1f4dac20ad8396b..47832f1cb0e44acbb30b560e8d6fd8d86e566010):
- Fixed dirty git tree ci failures
- Convert `PatchStatus` implementation to `Patch` with `subresources`
@xpivarc PTAL Thanks

**Comment by fossedihelm** [comment]
@Barakmor1 PTAL :slightly_smiling_face:

**Pull request merged**